orchagent/natorch.cpp

```diff
@@ -0,0 +1,4691 @@
+    snat_entry.data.mask.l4_src_port = 0xffff;
+    snat_entry.data.key.proto = ip_protocol;
+    snat_entry.data.mask.proto = 0xff;
+
+    status = sai_nat_api->remove_nat_entry(&snat_entry);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_INFO("Failed to removed %s SNAT NAPT entry with ip %s, port %d, prototype %s and it's translated ip %s, translated port %d",
+                      entry.entry_type.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port, keyEntry.prototype.c_str(),
+                      entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+    }
+    else
+    {
+        SWSS_LOG_NOTICE("Removed %s SNAT NAPT entry with ip %s, port %d, prototype %s and it's translated ip %s, translated port %d",
+                      entry.entry_type.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port, keyEntry.prototype.c_str(),
+                      entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+    }
+    deleteNaptCounters(keyEntry.prototype.c_str(), keyEntry.ip_address, keyEntry.l4_port);
+    m_naptEntries.erase(keyEntry);
+
+    if (entry.entry_type == "static")
+    {
+        if (totalStaticNaptEntries)
+        {
+            totalStaticNaptEntries--;
+            updateStaticNaptCounters(totalStaticNaptEntries);
+        }
+    }
+    else
+    {
+        if (totalDynamicNaptEntries)
+        {
+            totalDynamicNaptEntries--;
+            updateDynamicNaptCounters(totalDynamicNaptEntries);
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Found the total number dynamic napt entries to be corrupt, when removing SNAPT entry with proto %s, ip %s, port %d, translated ip %s, translated port %d!!",
+                            keyEntry.prototype.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port,
+                            entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+        }
+    }
+
+    if (totalSnatEntries)
+    {
+        totalSnatEntries--;
+        updateSnatCounters(totalSnatEntries);
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Found the total number dynamic snat entries to be corrupt, when removing SNAPT entry with proto %s, ip %s, port %d, translated ip %s, translated port %d!!",
+                       keyEntry.prototype.c_str(), keyEntry.ip_address.to_string().c_str(), keyEntry.l4_port,
+                       entry.translated_ip.to_string().c_str(), entry.translated_l4_port);
+    }
+
+    if (totalEntries)
+        totalEntries--;
+
+    return true;
+}
+
+bool NatOrch::addNatEntry(const IpAddress &ip_address, const NatEntryValue &entry)
+{
+    SWSS_LOG_ENTER();
+
+    /* Check the entry is present in cache */
+    if (m_natEntries.find(ip_address) != m_natEntries.end())
+    {
+        SWSS_LOG_INFO("Duplicate %s %s NAT entry with ip %s and it's translated ip %s, do nothing",
+                      entry.entry_type.c_str(), entry.nat_type.c_str(), ip_address.to_stri
```

Comment:
do we need to clean up the entries which are added to HW as well ?

Reply:
Yes, it will also delete the entries from hardware.